portal/views/reporting.py

```diff
@@ -223,105 +218,20 @@ def questionnaire_status():
-        last_viable = qb_stats.current_qbd(
-            even_if_withdrawn=True) or qb_stats.prev_qbd
-        if last_viable:
-            row['qb'] = last_viable.questionnaire_bank.name
-            row['visit'] = visit_name(last_viable)
-            entry_method = QNR_results(
-                patient, last_viable.qb_id,
-                last_viable.iteration).entry_method()
-            if entry_method:
-                row['entry_method'] = entry_method
-
-        results.append(row)
-
-        # as we require a full history, continue to add rows for each previous
-        # visit available
-        for qbd, status in qb_stats.older_qbds(last_viable):
-            historic = row.copy()
-            historic['status'] = status
-            historic['qb'] = qbd.questionnaire_bank.name
-            historic['visit'] = visit_name(qbd)
-            entry_method = QNR_results(
-                patient, qbd.qb_id, qbd.iteration).entry_method()
-            if entry_method:
-                historic['entry_method'] = entry_method
-            else:
-                historic.pop('entry_method', None)
-            results.append(historic)
-
-    if request.args.get('format', 'json').lower() == 'csv':
-        def gen(items):
-            desired_order = [
-                'user_id', 'study_id', 'status', 'visit', 'site', 'consent']
-            yield ','.join(desired_order) + '\n'  # header row
-            for i in items:
-                yield ','.join(
-                    ['"{}"'.format(i.get(k, "")) for k in desired_order]
-                ) + '\n'
-
-        # default file base title
-        base_name = 'Questionnaire-Timeline-Data'
-        if org_id:
-            base_name = '{}-{}'.format(
-                base_name,
-                Organization.query.get(org_id).name.replace(' ', '-'))
-        filename = '{}-{}.csv'.format(base_name, strftime('%Y_%m_%d-%H_%M'))
-
-        return Response(
-            gen(results),
-            headers={
-                'Content-Disposition': 'attachment;filename={}'.format(
-                    filename),
-                'Content-type': "text/csv"}
-        )
-    else:
-        return jsonify(bundle_results(elements=results))
+    from ..tasks import adherence_report_task
+
+    # This frequently takes over a minute to produce.  Generate a serializable
+    # form of all args for reliable hand off to a background task.
+    kwargs = {
+        'requested_as_of_date': request.args.get('as_of_date'),
+        'acting_user_id': current_user().id,
+        'include_test_role': request.args.get('include_test_role', False),
+        'org_id': request.args.get('org_id'),
+        'response_format': request.args.get('format', 'json').lower()
+    }
+
+    # Hand the task off to the job queue, and return 202 with URL for
+    # checking the status of the task
+    task = adherence_report_task.apply_async(kwargs=kwargs)
+    return jsonify({}), 202, {'Location': url_for(
```

Comment:
@achen2401 @pbugni 
I just wanted to add that I confirmed `url_for()` will return an `https` prefixed URL in prod deployments, without `scheme` needing to be explicitly passed.

just FYI I used this logging line to test `request.scheme` during a request
`current_app.logger.debug('scheme: %s', request.scheme)`